data/src/main/java/com/linkedin/data/schema/validation/ValidateDataAgainstSchema.java

```diff
@@ -142,6 +144,53 @@ protected void validate(DataElement element)
       {
         validateIterative(element);
       }
+
+      if (_toTrim.size() > 0) {
+        for (ToTrim toTrim: _toTrim) {
+          toTrim.trim();
+        }
+      }
+    }
+
+    /**
+     * Holds a data elements for later removal.
+     *
+     * This is a specialized version of {@link com.linkedin.data.it.Remover.ToRemove} which is and
+     * should remain private, so could not be used here.
+     */
+    private class ToTrim
+    {
+      private ToTrim(DataElement element)
+      {
+        DataElement parentElement = element.getParent();
+        _path = element.path();
+        _parent = parentElement == null ? null : parentElement.getValue();
+        _name = element.getName();
+      }
+
+      private void trim()
+      {
+        if (_parent != null) {
+          Class<?> nameClass = _name.getClass();
+          Class<?> parentClass = _parent.getClass();
+          if (nameClass == String.class) {
+            assert (parentClass == DataMap.class);
+            DataMap map = (DataMap) _parent;
+            if (map.isReadOnly())
+            {
+              _hasFixupReadOnlyError = true;
+              addMessage(_path, "unrecognized field cannot be trimmed because DataMap backing it is read-only");
+            }
+            else {
+              map.remove(_name);
+            }
+          }
+        }
+      }
+
+      private Object[] _path;
```

Comment:
Thinking about this diff more, I should remove this field, it's expensive to compute.